harness/determined/common/api/_session.py

```diff
@@ -22,69 +22,22 @@
 )
 
 
-def _do_request(
-    method: str,
-    host: str,
-    path: str,
-    max_retries: Optional[GeneralizedRetry],
-    params: Optional[Dict[str, Any]] = None,
-    json: Any = None,
-    data: Optional[str] = None,
-    headers: Optional[Dict[str, str]] = None,
-    cert: Optional[certs.Cert] = None,
-    timeout: Optional[Union[Tuple, float]] = None,
-    stream: bool = False,
-) -> requests.Response:
-    # Allow the json to come pre-encoded, if we need custom encoding.
-    if json is not None and data is not None:
-        raise ValueError("json and data must not be provided together")
-
-    if json:
-        data = det.util.json_encode(json)
-
-    relpath = path.lstrip("/")
-
-    try:
-        r = det_requests.request(
-            method,
-            f"{host}/{relpath}",
-            params=params,
-            data=data,
-            headers=headers,
-            verify=cert.bundle if cert else None,
-            stream=stream,
-            timeout=timeout,
-            server_hostname=cert.name if cert else None,
-            max_retries=max_retries,
-        )
-    except requests.exceptions.SSLError:
-        raise
-    except requests.exceptions.ConnectionError as e:
-        raise errors.MasterNotFoundException(str(e))
-    except requests.exceptions.RequestException as e:
-        raise errors.BadRequestException(str(e))
-
-    def _get_error_str(r: requests.models.Response) -> str:
-        try:
-            json_resp = _json.loads(r.text)
-            mes = json_resp.get("message")
-            if mes is not None:
-                return str(mes)
-            # Try getting GRPC error description if message does not exist.
-            return str(json_resp.get("error").get("error"))
-        except Exception:
-            return ""
-
-    if r.status_code == 403:
-        raise errors.ForbiddenException(message=_get_error_str(r))
-    if r.status_code == 401:
-        raise errors.UnauthenticatedException()
-    elif r.status_code == 404:
-        raise errors.NotFoundException(_get_error_str(r))
-    elif r.status_code >= 300:
-        raise errors.APIException(r)
-
-    return r
+def _make_requests_session(
+    server_hostname: Optional[str] = None,
+    verify: Optional[Union[str, bool]] = None,
```

Comment:
When `verify` is optional, then I have to figure out what three states mean. I think it's slightly easier (and means the same thing) if it's just a `Union[str, bool]` instead.

Reply:
we probably could default None to True, but it's not that obvious and i'm not sure that it's this method's responsibility to make that default decision.

`verify` comes directly from `cert.bundle`, which has `Optional[Union[str, bool]]`, along with some not-super-trivial logic to determine which type it is:
```
if noverify:
            self._bundle = False  # type: Union[None, str, bool]
        elif cert_pem is None:
            self._bundle = None
        else:
            self._bundle = str
```
 requests also takes in the same type and has some logic for handling None (ultimately it treats it as true AFAICT, but it's not immediately obvious).

Reply:
I'm only now looking at how `requests.Sessions` are initialized.

...

ok. So a `requests.Session` has a `verify` attribute that's set to `True` at init but which [the docs say](https://requests.readthedocs.io/en/latest/api/#requests.Session.verify) you can falsify later on. FWIW, I don't see mention of setting it to `None`.

`requests.Session.request` _does_ take an `Optional` `verify` when constructing its request, and that logic is _awful_:

1. if `request.verify \in (True | None)`, and a bundle-string exists in an environment variable, then overwrite `request.verify` with that bundle-string
2. if now `request.verify is None`, then use Session.verify for the request
3. if `request.verify is not None`, use whatever `request.verify` now is for the request

ok. It's time to send the request. Now it's up to the Adapter to decide what to do with `verify`. Looking at [the source](https://github.com/psf/requests/blob/f1bb07d39b74d6444e333879f8b8a3d9dd4d2311/src/requests/adapters.py#L124), the comment on the base class says that `verify` should be either a string or a bool. This library is untyped so whatever, but says nothing about the possibility that it's a `None`.

I think it's a type error at this point for `verify` to be a `None` unless the adapter we subclass it with explicitly handles that case. Ours does not, but having earlier set `request.verify` to `None` manually does mean that possibility is there.

My conclusion from this is that for our use case we should not set a `requests.Session.verify` to `None`.

Reply:
defaulting this to true now